### PR TITLE
Add token authorization option for extensions

### PR DESCRIPTION
**Pull request by jaimeteb**
Addresses #11 

Tests pending.

The token will only be supported by REST extensions.

@ryancurrah what do you think?

### CODE REVIEW

**Comment by ryancurrah** [comment]
This looks awesome. Should we do the same thing to the Chatto rest channel?

**Comment by jaimeteb** [comment]
> This looks awesome. Should we do the same thing to the Chatto rest channel in this PR?

Yeah! I didn't think about it but I'll do it, seems like a nice security addition.

**Comment by jaimeteb** [comment]
@ryancurrah 

How about this?

```yaml
telegram:
  bot_key:

twilio:
  account_sid:
  auth_token:
  number:

rest:
  token: my-super-secret-rest-token
```

**Comment by ryancurrah** [comment]
Looks good

**Comment by jaimeteb** [comment]
> Looks good

The token should be required in `/predict` and `/senders/{sender}` as well, right?

**Review by ryancurrah — COMMENTED** [review]
> > Looks good
> 
> The token should be required in `/predict` and `/senders/{sender}` as well, right?

Actually yeah that's a good point. It should probably be a `bot` api auth token. I guess the `Channel.REST` is weird one.

File: extension/server.go [hunk 1]
```diff
@@ -107,6 +117,7 @@ func (l *ListenerRPC) GetAllFuncs(req *Request, res *GetAllFuncsResponse) error
 // ListenerREST contains the RegisteredFuncs to be served through REST
 type ListenerREST struct {
 	RegisteredFuncs RegisteredFuncs
```
[the remark this answers is not in the document]
Comment (jaimeteb) [reply extension/server.go:109]: Yeah, let's use this PR.

**Comment by jaimeteb** [comment]
> Actually yeah that's a good point. It should probably be a `bot` api auth token for those.
> 
> Seems Slack and the other Channels support some sort of authentication in their own way.
> 
> Slack:
> https://api.slack.com/apis/connections/events-api#the-events-api__receiving-events__callback-field-overview
> 
> Twilio:
> https://www.twilio.com/docs/usage/webhooks/getting-started-twilio-webhooks#validate-that-webhook-requests-are-coming-from-twilio

That would be great for security. I'll start adding the `bot` token for `/channels/rest`, `/predict` and `/senders/{sender}`.

**Comment by ryancurrah** [comment]
Channel auth implementation is going to be different per channel client. So I would stick with your current implementation for rest. But also implement a bot auth for all bot endpoints that start with /bot/predict. That would be clean separation I think.

**Comment by jaimeteb** [comment]
Gotcha, so:

```yaml
# chn.yml
telegram:
  bot_key:

twilio:
  account_sid:
  auth_token:
  number:

rest:
  token: this-is-a-rest-token
```

On the other hand:

```yaml
extensions:
  type: REST
  url: http://localhost:8770

store:
  type: REDIS
  host: localhost
  password: pass
  ttl: 3600

auth:
  token: this-is-a-bot-token
```

**Comment by ryancurrah** [comment]
Beauty

**Comment by ryancurrah** [comment]
Sorry for rest would it make sense to distinguish between client and server auth?

**Comment by ryancurrah** [comment]
I say that because so far all the configuration in channels has been for setting up the channel client. Maybe separating those would be good?

channels:
  slack:
    callback_token: my-callback-token
    client:
      app_token: xpp-blah

**Comment by jaimeteb** [comment]
> I say that because so far all the configuration in channels has been for setting up the channel client. Maybe separating those would be good?
> 
> channels:
> slack:
> callback_token: my-callback-token
> client:
> app_token: xpp-blah

Not sure I follow. Could you format that small example?

**Comment by ryancurrah** [comment]
> > I say that because so far all the configuration in channels has been for setting up the channel client. Maybe separating those would be good?
> > channels:
> > slack:
> > callback_token: my-callback-token
> > client:
> > app_token: xpp-blah
> 
> Not sure I follow. Could you format that small example?

Sorry about that I was on mobile.

```yaml
rest:
  # Auth token to be used by REST client when calling back to Chatto
  callback_token: my-callback-token

slack:
  # Auth token to be used by Slack when calling back to Chatto
  callback_token: my-callback-token

  # For configuring the Slack API Client
  client:
    token: MY_SLACK_TOKEN
```

**Comment by jaimeteb** [comment]
@ryancurrah This is ready. At the moment only the REST channel will be able to check for Authorization. Slack and Twilio use HMAC and I've been having issues. This will be added in the future.

**Comment by ryancurrah** [comment]
Looks good to me

**Pull request merged**